src/transformers/modeling_utils.py

```diff
@@ -479,10 +490,8 @@ def __init__(self, config: PretrainedConfig, *inputs, **kwargs):
         # Save config and origin of the pretrained weights if given in model
         self.config = config
         self.name_or_path = config.name_or_path
-        if getattr(self.config, "gradient_checkpointing", False):
-            self.gradient_checkpointing_enable()
-            # Remove the attribute now that is has been consumed, so it's no saved in the config.
-            delattr(self.config, "gradient_checkpointing")
+        if self.supports_gradient_checkpointing:
+            self._gradient_checkpointing_hook = self.register_forward_pre_hook(gradient_checkpointing_hook)
```

Comment:
This is the culprit for the failure I reported here: https://github.com/huggingface/transformers/pull/14408#issuecomment-971004220

Reply:
removing it fixes the problem

Reply:
Does this mean that DeepSpeed does not support PyTorch hooks?

Reply:
it does and uses those extensively. Which perhaps is the cause of the problem if some hooks disagree or don't follow the prescribed instruction of not modifying certain things.

If you're not sure about the cause I can investigate it and report back what I find.

Reply:
Investigated: the issue is:

https://github.com/huggingface/transformers/blob/b567510cff606c9dd67cb7c56169bc596590e700/src/transformers/modeling_utils.py#L423

Reply:
so it looks like Deepspeed is just a harbinger here, and any other application that also uses hooks will trigger this issue.

It appears that what happens is that the hook is being removed from the dict that is being traversed one or more frames above.

I looked at what others did to solve this and they had to move the hook removal outside of the hook itself and into the `forward` when it's safe to remove it.

For some reason I can't reproduce this with pt-1.10, which means that they have reworked the loop that traverses the hooks dict to allow hooks to self-remove - probably using a copy to traverse the dict.

Reply:
possible fix: https://github.com/huggingface/transformers/pull/14427